src/bin/pg_autoctl/keeper.c

```diff
@@ -1115,6 +1120,259 @@ keeper_maintain_replication_slots(Keeper *keeper)
+		 *
+		 * - if we failed to connect to the monitor (network split, monitor is
+		 *   in maintenance or being restarted, etc): in that case just return
+		 *   false and have the main loop handle the situation
+		 *
+		 * - if we could connect to the monitor and then failed to check that
+		 *   the version of the monitor is the one we expect, then we're not
+		 *   compatible with this monitor and that's a different story.
+		 */
+		if (monitor->pgsql.status != PG_CONNECTION_OK)
+		{
+			return false;
+		}
+
+		/*
+		 * Okay we're not compatible with the current version of the
+		 * pgautofailover extension on the monitor. The most plausible scenario
+		 * is that the monitor got update: we're still running e.g. 1.4 and the
+		 * monitor is running 1.5.
+		 *
+		 * In that case we exit, and because the keeper node-active service is
+		 * RP_PERMANENT the supervisor is going to restart this process. The
+		 * restart happens with fork() and exec(), so it uses the current
+		 * version of pg_autoctl binary on disk, which with luck has been
+		 * updated to e.g. 1.5 too.
+		 *
+		 * TL;DR: just exit now, have the service restarted by the supervisor
+		 * with the expected version of pg_autoctl that matches the monitor's
+		 * extension version.
+		 */
+		exit(EXIT_CODE_MONITOR);
+	}
+
+	if (doInit)
+	{
+		PostgresSetup *pgSetup = &(postgres->postgresSetup);
+		uint64_t system_identifier = pgSetup->control.system_identifier;
+
+		if (!monitor_set_group_system_identifier(monitor,
+												 keeperState->current_group,
+												 system_identifier))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+	}
+
+	/* We used to output that in INFO every 5s, which is too much chatter */
+	log_debug("Calling node_active for node %s/%d/%d with current state: "
+			  "%s, "
+			  "PostgreSQL %s running, "
+			  "sync_state is \"%s\", "
+			  "current lsn is \"%s\".",
+			  config->formation,
+			  keeperState->current_node_id,
+			  keeperState->current_group,
+			  NodeStateToString(keeperState->current_role),
+			  reportPgIsRunning ? "is" : "is not",
+			  postgres->pgsrSyncState,
+			  postgres->currentLSN);
+
+	/* ensure we use the correct retry policy with the monitor */
+	(void) pgsql_set_main_loop_retry_policy(&(monitor->pgsql.retryPolicy));
+
+	/*
+	 * Report the current state to the monitor and get the assigned state.
+	 */
+	return monitor_node_active(monitor,
+							   config->formation,
+							   keeperState->current_node_id,
+							   keeperState->current_group,
+							   keeperState->current_role,
+							   reportPgIsRunning,
+							   postgres->postgresSetup.control.timeline_id,
+							   postgres->currentLSN,
+							   postgres->pgsrSyncState,
+							   assignedState);
+}
+
+
+/*
+ * keeper_node_has_been_dropped checks if the local node is being dropped or
+ * has been dropped already from the monitor.
+ */
+bool
+keeper_node_has_been_dropped(Keeper *keeper, bool *dropped)
```

Comment:
This function does a bit more than the comment and name suggest. If the node is currently in the process of being dropped (assigned drop, but currently some other state), this process is finished by this function. Let's at least change the comment and probably also the name.

Reply:
Good point, edited the comments and renamed the function `keeper_ensure_node_has_been_dropped`.